Tweak the code, comments and utilities.

lrhn opened the pull request:
Nothing here should change behavior or API, only
implementation and presentation.

Introduced the `Mutex` class because `Pool` seemed 
overqualified/overcomplicated for the job of being a single-lock mutex.

Added a `JsonConvertible` interface, just to have
an extension method on `Iterable<JsonConvertible>` 
for doing `[for (var v in iterable) v.toJson()]`.
Happened too often, so a good candidate for a
helper function.

Every change here is a suggestion, none of it is
essential. Feel free to ignore any parts that
seem unnecessary.
(The task-URI cache is probably unnecessary.)

lrhn commented on samples/dart/bin/util/resource.dart:
```diff
@@ -17,7 +17,16 @@ import 'dart:typed_data';
+/// Assumes the script is run from the `bin/` directory or another immediate
+/// subdirectory of the package root directory.
 Future<Uint8List> readResource(String name) {
-  final uri = Platform.script.resolve(path.join('..', 'resources', name));
```
This code would (maybe) not work on Windows.
The `path.join` would introduce backslashes, and `Uri.resolve` would not treat them as path separators.
Maybe `File.fromUri` would retain the `\`s, and then it worked anyway, but along the way, the URI was not a valid `file:` URI for the file.

(Never mix file paths and URIs. Always go between them explicitly using `Directory.fromUri(fileUro)` or `Uri.file(path)`, and only use `package:path` on file paths.)

devoncarew commented on pkgs/google_generative_ai/lib/google_generative_ai.dart:
```diff
@@ -12,12 +12,13 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+/// Google generative AI SDK API bindings for Dart.
```
As a drive-by-comment, if updating this text, can you also update the package description in the pubspec?

lrhn replied:
Done, PTAL

devoncarew commented on pkgs/google_generative_ai/pubspec.yaml:
```diff
@@ -11,7 +11,7 @@ environment:
 dependencies:
   http: ^1.1.0
-  pool: ^1.5.0
+  meta: ^1.11.0
```
This should address the CI failure:

```suggestion
  meta: ^1.10.0
```

lrhn committed:
Update pkgs/google_generative_ai/pubspec.yaml

Co-authored-by: Devon Carew <devoncarew@google.com>

natebosch commented on pkgs/google_generative_ai/lib/src/chat.dart:
```diff
@@ -88,61 +88,72 @@ final class ChatSession {
   /// Waits to read the entire streamed response before recording the message
-  /// and response and allow pending messages to be sent.
-  Stream<GenerateContentResponse> sendMessageStream(Content message) async* {
-    final resource = await _pool.request();
```
I think buffering it in another StreamController makes sense. This will also let us eagerly add it to the history even if the returned stream doesn't have a listener.
If we want to avoid some of the overhead of buffering in two places we could add to history and return a stream that can read from history when it gets a listener.

lrhn replied:
I don't think what we store in the history, and what's in the stream events, is necessarily the same thing.
The stream events are the original server events, which can (according to the class APIs, at least) include multiple candidates.
The content stored in the history is aggregated text from the *first* candidate of each event.

lrhn replied:
Anyway, I rewrote the function to use a stream controller to cache the events.
It was a remarkably simple rewrite.

natebosch commented on pkgs/google_generative_ai/lib/src/utils/json.dart:
```diff
@@ -0,0 +1,30 @@
+///
+/// Allows generalized helper functionality that works on different,
+/// otherwise unrelated, classes.
+abstract interface class JsonConvertible {
```
I was hoping to avoid exposing serialization details in the package public API. It was too noisy to drop a public `toJson` in my earlier versions, but after changing to a deep `toJson()` it's no longer relying on the behavior of that specific method in `jsonEncode`. I might look at pulling the serialization out into an extension or a private method.

lrhn replied:
ACK.
It really is an internal detail, only existing to allow the extension method on `List<JsonConvertible>`.

If we could make it a private class, `_JsonConvertible`, then it wouldn't leak, but every immediate subclass then needs to be declared in the same library.

lrhn committed:
Change `sendMessageStream` to start eagerly.

Now starts as as soon as possible, caching events in a
stream controller. This ensures it won't block other
requests, or get sent out of order, in case nobody
listens to the result stream.

lrhn committed:
Tweak the code, comments and utilities.

Nothing here should change behavior or API, only
implementation and presentation.

Introduced the `Mutex` class because `Pool` seemed
overqualified/overcomplicated for the job of being
a single-lock mutex.

Added a `JsonConvertible` interface, just to have
an extension method on `Iterable<JsonConvertible>`
for doing `[for (var v in iterable) v.toJson()]`.
Happened too often, so a good candidate for a
helper function.

Every change here is a suggestion, none of it is
essential. Feel free to ignore any parts that
seem unnecessary.
(The task-URI cache is probably unnecessary.)

lrhn committed:
Merge branch 'main' into tweak_code_lrhn

natebosch commented:
Closing since I don't think we plan on landing further changes from this branch.
